### PR TITLE
Finished Sell Land module + test

### PR DESCRIPTION
**Pull request by lisv**
User Stories: 
- Implement strategy for when and how to sell land
- Implement additional strategy for selling land when a reaction has been received
- Write tests for the features you're implementing this sprint

We only want to sell buildings that are not part of housing, as we do not need those. We first try to sell these multipolygons to the facilities stakeholder, for a maximum of three times, for which we reduce the selling price each time. 

If they still did not buy it we try selling it to other stakeholders in the same way.

### CODE REVIEW

**Comment by MatthijsJWH** [comment]
This module can (at least partially) and should be tested before it is merged in my opinion. Or can you give me a explanation as to why it can't be tested?

**Comment by lisv** [comment]
Okay, I'm working on it

**Comment by lisv** [comment]
Passing test has been added.

File: src/nl/codefox/contextvh/knowledgebase.pl [hunk 1]
```diff
@@ -23,50 +23,70 @@
...
 % predicates to determine if land on the map is ours or others
 landOfOthers(MultiPoly) :- stakeholder(StakeholderID,'Private_Woningbouw_Burgers',_,_),
 	not(land(LandID,stakeholder(StakeholderID,_,_,_),MultiPoly,_,_)),
 	MultiPoly\=multipolygon('MULTIPOLYGON EMPTY').
 ourLand(MultiPoly) :- stakeholder(StakeholderID,'Private_Woningbouw_Burgers',_,_),
 	land(LandID,stakeholder(StakeholderID,_,_,_),MultiPoly,_,_),
 	MultiPoly\=multipolygon('MULTIPOLYGON EMPTY').
+	
+% categories that are a member of housing
+housingMember(Categories) :- member('LUXE',Categories); member('NORMAL',Categories); member('SOCIAL',Categories); member('OTHER',Categories).
```
Comment (shdshdw) [inline src/nl/codefox/contextvh/knowledgebase.pl]: I think housingMember is a somewhat confusing name for this predicate.

File: src/nl/codefox/contextvh/sellLand.mod2g [hunk 2]
```diff
@@ -5,15 +5,33 @@ exit=always.
...
 % The selling land module
 module sellLand(MultiPoly) {
+	
+	% check if we succeeded in selling the land
+	if bel(not(ourLand(MultiPoly)),offeredLand(MultiPoly,BuyerID,Counter))
+		then delete(offeredLand(MultiPoly,BuyerID,Counter)) + 
+		insert(doneSelling(MultiPoly)) + exit-module.
+	
+	% try another buyer when we have tried it three times with facilities
+	if bel(offeredLand(MultiPoly,BuyerID,Counter),facilitiesID(BuyerID),Counter=3,
```
Comment (shdshdw) [inline src/nl/codefox/contextvh/sellLand.mod2g]: I believe it is better to handle this logic to adopt the goal to sell the correct land instead of checking that in this module. Mostly because it will then only execute the module when there really is a need to sell land.

**Comment by shdshdw** [comment]
When I run the code, there are no single occurrences where offeredLand(MultiPoly, Counter) is inserted with Counter > 1,and because there is no request handling for sell land I would like to know when does this counter actually become 2 or higher?
Edit: I ran this while Marco acted as the facilities stakeholder. After declining a sell land the counter does not get updated. 
On a sidenote: why do we actually sell about ALL our land?

**Comment by MHMCD** [comment]
So we decided to look what is happening in Tygron when the bot is running, we found quite some interesting interactions with the environment. Dare I say so, Terrifying.
As you can see in this beautiful picture:
https://i.gyazo.com/2d51b23706c808387d1b7f37956227da.png
We are currently selling all the gardens around our owned buildings, this is not what I meant with selling certain facilities. Now I understand that you did not take into account that these would be brought up to sell but I would expect of you to at least check in the tygron engine what is happening when your bot is running with your new code, you told us that your tygron engine is not working, then ask us to run it for you so we can give you feedback on what your bot is doing. The counter does not work at the moment, you are selling way too much ground and you believe the code is ready to be merged into the development branch? Could you please at least look at what your bot is doing before you try to pull request your branch.

**Comment by lisv** [comment]
You were right, this branch was not ready to be pulled into the developing branch.

However, I have rewritten it and manually tested it (with Marco) and now it works correctly:
- The sellLand module is tested
- We only sell buildings that are not part of housing now
- The counter works, using the requestsmodule
- The code has been manually tested

Please review it again.

**Comment by shdshdw** [comment]
Resolve the conflicts please.

**Comment by lisv** [comment]
Conflicts are currently resolved

File: src/nl/codefox/contextvh/knowledgebase.pl [hunk 3]
```diff
@@ -52,31 +50,57 @@
...
 	indicator(ID,_,_,ZoneLinkList),member(zone_link(1,_,Current2,Target2),ZoneLinkList),
 	Target1-Current1>Target2-Current2.
-
+% indicator of the spatial quality
+qualityIndicator(Value,Target,ZoneID) :-
+	indicatorLink(_,IndicatorWeights), member(indicatorWeights(IndicatorID,IndicatorName,_),IndicatorWeights),
+	(IndicatorName=='Ruimtelijke kwaliteit'), indicator(IndicatorID,Value,Target,ZoneLink),
+	member(zone_link(ZoneID,IndicatorID,Value,Target),ZoneLink).
+% indicator of the sound
+soundIndicator(Value,Target,ZoneID) :-
```
Comment (shdshdw) [inline src/nl/codefox/contextvh/knowledgebase.pl:103]: Although I think it is good to use predicates like this instead of hard-coding indicatorId's, I think this is a change that is outside of the scope of this PR. Please consider this in next PR's.

**Comment by lisv** [comment]
Also, a time out should also be implemented so that the agent will move on if it does not receive a reaction back from a stakeholder but this is too big of a task for now. We should do this soon in a different branch.

File: src/nl/codefox/contextvh/tygronGoals.mod2g [hunk 4]
```diff
@@ -4,6 +4,17 @@ order = linearall.
 module tygronGoals {
 	%_______________________________Adopt goals based on the environment_______________________________%
 	
+	%%% Selling buildings
+	
+	% If we want to sell a building but failed to do it before, we demolish it.
+	if bel(building(BuildingID,Name,OwnerID,_,Categories,FunctionID,_,MultiPoly,_,_),ourID(OwnerID),not(housingBuildings(Categories)),
+		failedToSell(MultiPoly))
```
Comment (shdshdw) [inline src/nl/codefox/contextvh/tygronGoals.mod2g:11]: As this functionality has not yet been triggered, please check if this works correctly. Because this functionality does not break other code I agree to do so after the merge.

**Comment by shdshdw** [comment]
Approved, but please extend this after the merge.

**Comment by MatthijsJWH** [comment]
I approve, nicely done adding the test

**Comment by MHMCD** [comment]
I approve, we have done multiple tests and I believe we ruled out the bugs we were experiencing earlier.

File: lib/tygron [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit 41169d27e37352fc367d4bdfbfa9cf7e3480a1f6
```
Comment (JoshCode) [inline lib/tygron]: This is the wrong submodule commit

**Comment by JoshCode** [comment]
Looks ready to be merged now.

File: lib/tygron (hunk 5 above)
Comment (lisv) [inline lib/tygron]: Fixed

**Comment by JoshCode** [comment]
Travis builds should be passing, it's a problem with Travis, not our bot. Merging.
